truffleruby/src/main/java/org/truffleruby/core/thread/ThreadNodes.java

```diff
@@ -415,15 +415,9 @@ public DynamicObject wakeup(DynamicObject thread,
 
         @Specialization(guards = {"isRubyProc(unblocker)", "isRubyProc(runner)"})
         public Object unblock(VirtualFrame frame, DynamicObject thread, DynamicObject unblocker, DynamicObject runner) {
-            Layouts.THREAD.setUnblocker(thread, unblocker);
-            Layouts.THREAD.setStatus(thread, ThreadStatus.SLEEP);
-
-            try {
-                return yield(frame, runner);
-            } finally {
-                Layouts.THREAD.setUnblocker(thread, null);
-                Layouts.THREAD.setStatus(thread, ThreadStatus.RUN);
-            }
+            return getContext().getThreadManager().runUntilResult(this,
+                    () -> yield(frame, runner),
+                    () -> yield(null, unblocker));
```

Comment:
I matched the code I was replacing. I assumed it was originally passed `null` for a reason.